### DISCUSSION TITLE
Code after yield in the async dependecy executed before response has been sent to the client

### DISCUSSION THREAD

**Discussion by FNSdev** [fastapi/fastapi#8529] [Questions]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
import asyncio
from typing import AsyncGenerator

import structlog
import uvicorn
from fastapi import FastAPI, Depends
from starlette.middleware.base import BaseHTTPMiddleware, RequestResponseEndpoint
from starlette.requests import Request
from starlette.responses import Response

logger = structlog.getLogger(__name__)


app = FastAPI()


class BarMiddleware(BaseHTTPMiddleware):
    async def dispatch(self, request: Request, call_next: RequestResponseEndpoint) -> Response:
        logger.info("BAR: Before call_next()")
        response = await call_next(request)
        logger.info("BAR: After call_next()")
        logger.info(f"BAR: Status code {response.status_code}")

        logger.info("BAR: Before sleep()")
        await asyncio.sleep(2)
        logger.info("BAR: After sleep()")

        return response


class ZxcMiddleware(BaseHTTPMiddleware):
    async def dispatch(self, request: Request, call_next: RequestResponseEndpoint) -> Response:
        logger.info("ZXC: Before call_next()")
        response = await call_next(request)
        logger.info("ZXC: After call_next()")
        logger.info(f"ZXC: Status code {response.status_code}")

        logger.info("ZXC: Before sleep()")
        await asyncio.sleep(2)
        logger.info("ZXC: After sleep()")

        return response


async def get_foo() -> AsyncGenerator[str, None]:
    logger.info("BEFORE YIELD FOO")
    yield "foo"
    logger.info("AFTER YIELD FOO")


async def get_qwerty(foo: str = Depends(get_foo)) -> AsyncGenerator[str, None]:
    logger.info("BEFORE YIELD QWERTY")
    yield "qwerty"
    logger.info("AFTER YIELD QWERTY")


@app.get("/")
async def root(qwerty: str = Depends(get_qwerty)):
    return {"message": qwerty}


app.add_middleware(BarMiddleware)
app.add_middleware(ZxcMiddleware)


if __name__ == "__main__":
    uvicorn.run(
        "test:app",
        host="0.0.0.0",
        port=8080,
        reload=True,
    )
```

### Description

Hi.

From the documentation:
> The code following the yield statement is executed after the response has been delivered:

But this is not the case:
```
2022-06-23 09:21:29,932 uvicorn.error INFO: Finished server process [30197]
2022-06-23 09:21:30,309 uvicorn.error INFO: Started server process [30216]
2022-06-23 09:21:30,309 uvicorn.error INFO: Waiting for application startup.
2022-06-23 09:21:30,309 uvicorn.error INFO: Application startup complete.
2022-06-23 09:22.12 [info     ] ZXC: Before call_next()
2022-06-23 09:22.12 [info     ] BAR: Before call_next()
2022-06-23 09:22.12 [info     ] BEFORE YIELD FOO
2022-06-23 09:22.12 [info     ] BEFORE YIELD QWERTY
2022-06-23 09:22.12 [info     ] BAR: After call_next()
2022-06-23 09:22.12 [info     ] BAR: Status code 200
2022-06-23 09:22.12 [info     ] BAR: Before sleep()
2022-06-23 09:22.14 [info     ] BAR: After sleep()
2022-06-23 09:22.14 [info     ] ZXC: After call_next()
2022-06-23 09:22.14 [info     ] ZXC: Status code 200
2022-06-23 09:22.14 [info     ] ZXC: Before sleep()
2022-06-23 09:22.14 [info     ] AFTER YIELD QWERTY
2022-06-23 09:22.14 [info     ] AFTER YIELD FOO
2022-06-23 09:22.16 [info     ] ZXC: After sleep()
2022-06-23 09:22:16,494 uvicorn.access INFO: 127.0.0.1:55382 - "GET / HTTP/1.1" 200
```

Response has not been sent to the client yet, but we're already in the "after yield" code. I can not reproduce it with a single middleware though.

Using native ASGI middlewares (instead of inheriting from BaseHTTPMiddleware) does not solve the issue.

It's a problem for us, because we can not rely on the fact that transaction has been commited already (in the middleware), when we're in the "after yield" code.

Either docs are misleading, or something works incorrectly.

### Operating System

Linux

### FastAPI Version

0.78.0

### Python Version

3.8.10

**Comment by JarroVGIT** [comment]
Could you please post your code in which you tried native ASGI middleware? You say you tried it, but didn't solve the issue, can we assume that it yielded the same behaviour as with BaseHTTPMiddleware? 

I would highly recommend to use native ASGI middleware anyway, as BaseHTTPMiddleware ~~will~~ **might** be deprecated (see this open issue on the Starlette project: [Issue](https://github.com/encode/starlette/issues/1678)). So having the starting point where you see this behaviour with native ASGI middleware would help me reproduce this (and hopefully, help you fix it ;))

**Comment by JarroVGIT** [comment]
This is actually much stranger behaviour than I would expect to be honest. I wrote the ASGI middleware myself, and then the code after yield is even immediately called after the response has been sent (and before it has been handled by the middleware). I thought "maybe it is the asyncio.sleep, which is throwing back control to the event loop" but that made no difference at all. 

Code:
```python
import asyncio
from typing import AsyncGenerator

import uvicorn
import logging
from fastapi import FastAPI, Depends
from starlette.requests import Request
from starlette.responses import Response

logger = logging.getLogger()

app = FastAPI(debug=True)

class BarMiddleware:
    def __init__(self, app):
        self.app = app

    async def __call__(self, scope, receive, send):
        if scope["type"] != "http":
            return await self.app(scope, receive, send)
        logger.warning("BAR: Before call self.app()")   
        status_code = 1
        async def send_wrapper(message):
            if message["type"] == "http.response.start":
                nonlocal status_code
                status_code = message["status"]
            await send(message)
        r = await self.app(scope, receive, send_wrapper)
        logger.warning("BAR: After call self.app()")
        logger.warning(f"BAR: Status code {status_code}")
        logger.warning("BAR: Before sleep()")
        #await asyncio.sleep(2)
        logger.warning("BAR: After sleep()")

class ZxcMiddleware:
    def __init__(self, app):
        self.app = app

    async def __call__(self, scope, receive, send):
        if scope["type"] != "http":
            return await self.app(scope, receive, send)
        logger.warning("ZXC: Before call self.app()")
        status_code = 1
        async def send_wrapper(message):
            if message["type"] == "http.response.start":
                nonlocal status_code
                status_code = message["status"]
            await send(message)
        r = await self.app(scope, receive, send_wrapper)
        logger.warning("ZXC: After call self.app()")
        logger.warning(f"ZXC: Status code {status_code}")
        logger.warning("ZXC: Before sleep()")
        #await asyncio.sleep(2)
        logger.warning("ZXC: After sleep()")

async def get_foo() -> AsyncGenerator[str, None]:
    logger.warning("BEFORE YIELD FOO")
    yield "foo"
    logger.warning("AFTER YIELD FOO")

async def get_qwerty(foo: str = Depends(get_foo)) -> AsyncGenerator[str, None]:
    logger.warning("BEFORE YIELD QWERTY")
    yield "qwerty"
    logger.warning("AFTER YIELD QWERTY")

app.add_middleware(BarMiddleware)
app.add_middleware(ZxcMiddleware)

@app.get("/")
async def root(qwerty: str = Depends(get_qwerty)):
    return {"message": qwerty}

if __name__ == "__main__":
    uvicorn.run(app, host="0.0.0.0", port=8000,  )
```

My logging:
```
ZXC: Before call self.app()
BAR: Before call self.app()
BEFORE YIELD FOO
BEFORE YIELD QWERTY
INFO:     127.0.0.1:62160 - "GET / HTTP/1.1" 200 OK
AFTER YIELD QWERTY
AFTER YIELD FOO
BAR: After call self.app()
BAR: Status code 200
BAR: Before sleep()
BAR: After sleep()
ZXC: After call self.app()
ZXC: Status code 200
ZXC: Before sleep()
ZXC: After sleep()
```

This doesn't make sense to me at all to be honest. I am really curious to your ASGI middleware, and the behaviour in that.

**Comment by FNSdev** [comment]
That's what I have tried.

```
import asyncio
from typing import AsyncGenerator

import structlog
import uvicorn
from fastapi import FastAPI, Depends
from starlette.middleware.base import BaseHTTPMiddleware, RequestResponseEndpoint
from starlette.requests import Request
from starlette.responses import Response

logger = structlog.getLogger(__name__)


app = FastAPI()


@app.middleware("http")
async def bar_middleware(request, call_next):
    logger.info("BAR: Before call_next()")
    response = await call_next(request)
    logger.info("BAR: After call_next()")

    logger.info("BAR: Before sleep()")
    await asyncio.sleep(2)
    logger.info("BAR: After sleep()")

    return response


@app.middleware("http")
async def zxc_middleware(request, call_next):
    logger.info("ZXC: Before call_next()")
    response = await call_next(request)
    logger.info("ZXC: After call_next()")
    logger.info(f"ZXC: Status code {response.status_code}")

    logger.info("ZXC: Before sleep()")
    await asyncio.sleep(2)
    logger.info("ZXC: After sleep()")

    return response


async def get_foo() -> AsyncGenerator[str, None]:
    logger.info("BEFORE YIELD FOO")
    yield "foo"
    logger.info("AFTER YIELD FOO")


async def get_qwerty(foo: str = Depends(get_foo)) -> AsyncGenerator[str, None]:
    logger.info("BEFORE YIELD QWERTY")
    yield "qwerty"
    logger.info("AFTER YIELD QWERTY")


@app.get("/")
async def root(qwerty: str = Depends(get_qwerty)):
    return {"message": qwerty}


if __name__ == "__main__":
    uvicorn.run(
        "test:app",
        host="0.0.0.0",
        port=8080,
        reload=True,
    )
```

The result:
![image](https://user-images.githubusercontent.com/29799729/175472780-37b2dc05-dc0d-419e-bcf6-f187f315137e.png)

**Comment by JarroVGIT** [comment]
That is using httpbasemiddleware as well? (The decorator ‘ @app.middleware("http")’ is just syntax sugarcoating for inheriting from BaseHTTPMiddleware.

**Comment by FNSdev** [comment]
My bad, sorry, you're right. It works as I would expect with a real native middleware. Thank you.

```
import asyncio
from typing import AsyncGenerator

import uvicorn
import logging
from fastapi import FastAPI, Depends

logger = logging.getLogger()

app = FastAPI(debug=True)


class BarMiddleware:
    def __init__(self, app):
        self.app = app

    async def __call__(self, scope, receive, send):
        if scope["type"] != "http":
            return await self.app(scope, receive, send)
        logger.warning("BAR: Before call self.app()")

        async def send_wrapper(message):
            if message["type"] == "http.response.start":
                logger.warning("BAR: Before sleep()")
                await asyncio.sleep(2)
                logger.warning("BAR: After sleep()")

            await send(message)

        await self.app(scope, receive, send_wrapper)

        logger.warning("BAR: After call self.app()")


class ZxcMiddleware:
    def __init__(self, app):
        self.app = app

    async def __call__(self, scope, receive, send):
        if scope["type"] != "http":
            return await self.app(scope, receive, send)
        logger.warning("ZXC: Before call self.app()")

        async def send_wrapper(message):
            if message["type"] == "http.response.start":
                logger.warning("ZXC: Before sleep()")
                await asyncio.sleep(2)
                logger.warning("ZXC: After sleep()")

            await send(message)

        await self.app(scope, receive, send_wrapper)

        logger.warning("ZXC: After call self.app()")


async def get_foo() -> AsyncGenerator[str, None]:
    logger.warning("BEFORE YIELD FOO")
    yield "foo"
    logger.warning("AFTER YIELD FOO")


async def get_qwerty(foo: str = Depends(get_foo)) -> AsyncGenerator[str, None]:
    logger.warning("BEFORE YIELD QWERTY")
    yield "qwerty"
    logger.warning("AFTER YIELD QWERTY")


app.add_middleware(BarMiddleware)
app.add_middleware(ZxcMiddleware)


@app.get("/")
async def root(qwerty: str = Depends(get_qwerty)):
    return {"message": qwerty}


if __name__ == "__main__":
    uvicorn.run(
        "test:app",
        host="0.0.0.0",
        port=8080,
        reload=True,
    )

```

```
ZXC: Before call self.app()
BAR: Before call self.app()
BEFORE YIELD FOO
BEFORE YIELD QWERTY
BAR: Before sleep()
BAR: After sleep()
ZXC: Before sleep()
ZXC: After sleep()
INFO:     127.0.0.1:32962 - "GET / HTTP/1.1" 200 OK
AFTER YIELD QWERTY
AFTER YIELD FOO
BAR: After call self.app()
ZXC: After call self.app()
```

But I still believe that either docs need to be updated, or there must be a disclaimer to avoid using BaseHttpMiddleware.

**Comment by JarroVGIT** [comment]
Interesting, I thought that was the wrong order. The docs say:

https://github.com/tiangolo/fastapi/blob/1876ebc77949a9a254909ec61ea0c09365169ec2/docs/en/docs/tutorial/middleware.md?plain=1#L15

So, I figured the

```
AFTER YIELD QWERTY
AFTER YIELD FOO
```
would appear as latest in the logging stack. Maybe I am interpreting the documentation incorrectly, because the code after `yield` is being executed after the middleware has had its turn with the **incoming request**, but I expected the code dealing with **outgoing** response would adhere to this too (e.g. first you would see `BAR: After call self.app()`, then `ZXC: After call self.app()` and **then** `AFTER YIELD QWERTY/FOO`). 

By the way, that is why my response was that it was strange behaviour, funny that you expected that very behaviour your self 😋

**Comment by FNSdev** [comment]
```
INFO:     127.0.0.1:32962 - "GET / HTTP/1.1" 200 OK
AFTER YIELD QWERTY
AFTER YIELD FOO
```

So from my understanding, it looks correct, response was sent already, we got back to our dependencies.

The difference here is that we're actually logging in the middleware AFTER the response has been sent, which was not possible using BaseHttpMiddleware. But I don't see a problem here. Docs do not specify when we should get to that point, before or after deps.

In our case we're only interested in commiting / rolling back the transaction. It can be done in the place of "After sleep" log message (in the `send_wrapper`), which is executed before response has been sent. So it's exactly what we want.

**Comment by Kludex** [comment]
> BaseHTTPMiddleware will be deprecated (see this open issue on the Starlette project: [Issue](https://github.com/encode/starlette/issues/1678)). So having the starting point where you see this behaviour with native ASGI middleware would help me reproduce this (and hopefully, help you fix it ;))

That's not for sure yet... There has been a lot of discussion about `BaseHTTPMiddleware`. Also, the [limitations of `BaseHTTPMiddleware`](https://www.starlette.io/middleware/#basehttpmiddleware) are usually not a problem for most of the users.

That being said... They are still limitations, and if you are going to have them, for the time being, pure ASGI middlewares are recommended.

**Comment by JarroVGIT** [comment]
> > BaseHTTPMiddleware will be deprecated (see this open issue on the Starlette project: [Issue](https://github.com/encode/starlette/issues/1678)). So having the starting point where you see this behaviour with native ASGI middleware would help me reproduce this (and hopefully, help you fix it ;))
> 
> That's not for sure yet... There has been a lot of discussion about `BaseHTTPMiddleware`. Also, the [limitations of `BaseHTTPMiddleware`](https://www.starlette.io/middleware/#basehttpmiddleware) are usually not a problem for most of the users.
> 
> That being said... They are still limitations, and if you are going to have them, for the time being, pure ASGI middlewares are recommended.

Fair point, I corrected my comment above 😇

**Comment by YuriiMotov** [comment]
In current version the order of prints is different:
```
ZXC: Before call_next()
BAR: Before call_next()
BEFORE YIELD FOO
BEFORE YIELD QWERTY
AFTER YIELD QWERTY
AFTER YIELD FOO
BAR: After call_next()
BAR: Status code 200
BAR: Before sleep()
BAR: After sleep()
ZXC: After call_next()
ZXC: Status code 200
ZXC: Before sleep()
ZXC: After sleep()
INFO:     127.0.0.1:41780 - "GET / HTTP/1.1" 200 OK
```

And the order of dependencies with `yield` execution is described in docs as: 

> The code following the yield statement is executed after creating the response but before sending it

https://fastapi.tiangolo.com/tutorial/dependencies/dependencies-with-yield/#a-database-dependency-with-yield
